src/pieces.py

```diff
@@ -9,118 +9,272 @@
+        if not self.has_moved:
+            squares.add(square + (self.pawn_step + self.pawn_step) * self.color.value)
+        return squares
+
 
 @dataclass
-class Bishop(Piece):
-    """A bishop."""
+class Melee(Piece):
+    """A close ranged piece:
+        - King
+        - Knight (has reach)
 
-    value: int = 3
+    Pawns are special.
+    """
 
-    def __repr__(self) -> str:
-        """Represent the bishop.
+    def legal_moves(self, square: Square, condition: Callable[[Square], bool]) -> set[Square]:
+        super().legal_moves.__doc__
+        squares = set()
+        for move in self.steps:
+            if condition(square + move):
+                squares.add(square + move)
+        return squares
 
-        Returns:
-            The representation of a bishop.
-        """
-        return "♝" if self.is_black else "♗"
+
+@dataclass
+class King(Melee):
+    """A King.
+
+    Moves one square in any direction that is not blocked by same side piece or targeted by enemy piece.
+    """
+
+    steps = Melee.steps.union(Piece.straights.union(Piece.diagonals))
+
+    def __repr__(self) -> str:
+        super().__repr__.__doc__
+        return "♚" if self.is_black else "♔"
 
 
 @dataclass
-class Knight(Piece):
-    """A knight."""
+class Knight(Melee):
+    """A knight.
+
+    Moves two squares in any direction and one square in a vertical to that direction as one move.
+    It therefore can jump over other pieces. Target square must not be blocked by a same side piece.
+    """
+
+#   Knight moves
+    steps = Melee.steps.union(diagonal + straight for diagonal, straight in product(Piece.diagonals, Piece.straights))
+    steps = steps.difference(Piece.rizontals.union(Piece.straights))  # remove the non-knight moves
 
     value: int = 3
 
     def __repr__(self) -> str:
-        """Represent the knight.
-
-        Returns:
-            The representation of a knight.
-        """
+        super().__repr__.__doc__
         return "♞" if self.is_black else "♘"
 
 
 @dataclass
-class Rook(Piece):
-    """A rook."""
+class Range(Piece):
+    """A long range piece:
+        - Rook
+        - Bishop
+        - Queen
+    """
+
+    def legal_moves(self, square: Square, condition: Callable[[Square], bool]) -> set[Square]:
+        super().legal_moves.__doc__
+        squares = set()
+        for move in self.steps:  # For each direction.
+            advanced_square = square + move  # Start looking at advanced squares
+            while condition(advanced_square):  # While we don't hit something.
+                squares.add(advanced_square)  # Add the damn square to legal destination squares.
+                advanced_square += move  # Advance to the next square in line.
+        return squares
+
+
+@dataclass
+class Rook(Range):
+    """A rook.
+
+    Moves anywhere along its rank xor its file, unless blocked by enemy (which captures) or same piece.
+    """
+
+#   Straight lines:
+    steps = Range.steps.union(Piece.straights)
```

Comment:
```py
steps = set(Piece.straights)
```
(To create a separate object than `Piece.straights`)

Reply:
I imagined this to be unnecessary since sets are hashable, I'll take a look.